go/nbs/s3_table_persister.go

```diff
@@ -184,3 +194,180 @@ func (s partsByPartNum) Less(i, j int) bool {
 func (s partsByPartNum) Swap(i, j int) {
 	s[i], s[j] = s[j], s[i]
 }
+
+func (s3p s3TablePersister) CompactAll(sources chunkSources) chunkSource {
+	plan := planCompaction(sources)
+	if plan.chunkCount == 0 {
+		return emptyChunkSource{}
+	}
+	t1 := time.Now()
+	name := nameFromSuffixes(plan.suffixes())
+	s3p.multipartCopyUpload(plan, name.String())
+	verbose.Log("Compacted table of %d Kb in %s", plan.totalCompressedData/1024, time.Since(t1))
+
+	return s3p.newReaderFromIdxData(plan.mergedIndex, name)
+}
+
+func (s3p s3TablePersister) multipartCopyUpload(plan compactionPlan, key string) {
+	uploadID := s3p.startMultipartUpload(key)
+	multipartUpload, err := s3p.uploadPartsCopy(plan, key, uploadID)
+	if err != nil {
+		s3p.abortMultipartUpload(key, uploadID)
+		d.PanicIfError(err) // TODO: Better error handling here
+	}
+	s3p.completeMultipartUpload(key, uploadID, multipartUpload)
+}
+
+func (s3p s3TablePersister) uploadPartsCopy(plan compactionPlan, key, uploadID string) (*s3.CompletedMultipartUpload, error) {
+	d.PanicIfTrue(len(plan.sources) > maxS3Parts) // TODO: BUG 3433: handle > 10k parts
+
+	copies, manuals, buff := dividePlan(plan.sources, plan)
+	var readWg sync.WaitGroup
+	for _, man := range manuals {
+		readWg.Add(1)
+		go func(m manualPart) {
+			defer readWg.Done()
+			n, _ := m.r.Read(buff[m.buffStart:m.buffEnd])
+			d.PanicIfTrue(int64(n) < m.buffEnd-m.buffStart)
+		}(man)
+	}
+	readWg.Wait()
+
+	type uploadFn func() (etag string, err error)
+	sent, failed, done := make(chan s3UploadedPart), make(chan error), make(chan struct{})
+	var uploadWg sync.WaitGroup
+	sendPart := func(partNum int64, doUpload uploadFn) {
+		defer uploadWg.Done()
+
+		// Check if upload has been terminated
+		select {
+		case <-done:
+			return
+		default:
+		}
+
+		etag, err := doUpload()
+		if err != nil {
+			failed <- err
+			return
+		}
+		// Try to send along part info. In the case that the upload was aborted, reading from done allows this worker to exit correctly.
+		select {
+		case sent <- s3UploadedPart{int64(partNum), etag}:
+		case <-done:
+			return
+		}
+	}
+
+	numCopyParts := len(copies)
+	numManualParts := getNumParts(len(buff), s3p.partSize) // TODO: What if this is too big?
+	numParts := numCopyParts + numManualParts
```

Comment:
can we increment numParts and add to the wg for each call to go sendPart. The code below is linear, I think it'd be easier to follow, and you can just initialize nextPart := 1